command_definitions.py

```diff
@@ -0,0 +1,128 @@
+        if self.predicate():
+            return run(' '.join(commands))
+
+
+class Docker(CallList):
+    def __init__(self, command, predicate=lambda: True):
+        super(Docker, self).__init__(command=command, predicate=predicate)
+        self.base_command = ['docker']
+
+
+class Compose(CallList):
+    def __init__(self, command, predicate=lambda: True):
+        super(Compose, self).__init__(command=command, predicate=predicate)
+
+        self.base_command = [DOCKER_COMPOSE_EXECUTABLE]
+
+    def execute(self, project, hostname, environment=None):
+
+        provide_docker_compose()
+        compose_main_file = '{0}/docker-compose.yml'.format(project)
+        compose_host_file = '{0}/{1}.yml'.format(project, hostname)
+
+        mkdir(project)
+
+        put(local_path=compose_main_file, remote_path=compose_main_file)
+        put(local_path=compose_host_file, remote_path=compose_host_file)
+        env_file_path = "{}/*.env".format(project)
+        if len(glob.glob(env_file_path)) > 0:
+            put(local_path="{}/*.env".format(project), remote_path="{}/".format(project))
+
+        self.base_command += [
+            '-f', compose_main_file, '-f', compose_host_file
+        ]
+        super(Compose, self).execute(project, hostname, environment=environment)
+
+
+def is_running(container_name, name_is_regex=False):
+    if not name_is_regex:
+        container_name = re.escape(container_name)
+    result = run('docker ps')
+    m = re.search(r'\ {}\r?$'.format(container_name), result, flags=re.MULTILINE)
+    return m is not None
+
+
+def is_not_running(container_name, name_is_regex=False):
+    return not is_running(container_name, name_is_regex=name_is_regex)
+
+
+def has_succeeded(container_name, name_is_regex=False):
+    if not name_is_regex:
+        container_name = re.escape(container_name)
+    result = run('docker ps -a')
+    pattern = r'{status}.*{container_name}\r?$'.format(status=re.escape('Exited (0)'), container_name=container_name)
+    m = re.search(pattern, result, flags=re.MULTILINE)
+    return m is not None
+
+
+def has_not_succeeded(container_name, name_is_regex=False):
+    return not has_succeeded(container_name, name_is_regex=name_is_regex)
+
+
+def container_exists(container_name, name_is_regex=False):
+    if not name_is_regex:
+        container_name = re.escape(container_name)
+    result = run('docker ps -a')
+    pattern = r'.*{container_name}\r?$'.format(container_name=container_name)
+    m = re.search(pattern, result, flags=re.MULTILINE)
+    return m is not None
+
+
+def container_not_exists(container_name, name_is_regex=False):
+    return not container_exists(container_name, name_is_regex=name_is_regex)
+
+
+def has_succeeded_or_is_running(container_name, name_is_regex=False):
+    return is_running(container_name, name_is_regex=name_is_regex) or has_succeeded(container_name, name_is_regex=name_is_regex)
+
+
+def not_has_succeeded_or_is_running(container_name, name_is_regex=False):
```

Comment:
This name is prone to misinterpretation. With the usual operator precedence in boolean logic (and also [Python's operator precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence)), this would be interpreted as
> (NOT  has_succeeded) OR is_running

while you mean
> NOT (has_succeeded OR is_running)

which, without braces, would have to be written as
> NOT has_succeeded AND NOT is_running


Name suggestions (neither of which I'm really happy with):
* `has_not_succeeded_and_is_not_running`
* `neither_has_succeeded_nor_is_running`

Trying to think of a name without negation, but the best I can come up with is `needs_running_to_be_successful` which isn't less bewildering.